apps/sms/views/shared/js/activity_handler.js

```diff
@@ -21,63 +23,42 @@ const ActivityDataType = {
 };
 
 var ActivityHandler = {
-  // Will hold current activity object
-  _activity: null,
-
   init: function() {
     if (!window.navigator.mozSetMessageHandler) {
       return;
     }
 
-    // A mapping of MozActivity names to their associated event handler
-    window.navigator.mozSetMessageHandler('activity',
-      this._onActivity.bind(this, {
-        'new': this._onNewActivity,
-        'share': this._onShareActivity
-      })
-    );
-
-    // We don't want to register these system handlers when app is run as
-    // inline activity
-    if (!Navigation.getPanelName().startsWith('activity')) {
-      window.navigator.mozSetMessageHandler('sms-received',
-        this.onSmsReceived.bind(this));
-
-      window.navigator.mozSetMessageHandler('notification',
-        this.onNotification.bind(this));
-    }
-  },
-
-  isInActivity: function isInActivity() {
-    return !!this._activity;
-  },
-
-  setActivity: function setActivity(value) {
-    if (!value) {
-      throw new Error('Activity should be defined!');
-    }
-    this._activity = value;
-  },
+    if (ActivityService.hasPendingRequest()) {
+      ActivityClient.init();
 
-  // The Messaging application's global Activity handler. Delegates to specific
-  // handler based on the Activity name.
-  _onActivity: function activityHandler(handlers, activity) {
-    var name = activity.source.name;
-    var handler = handlers[name];
+      ActivityClient.on(
+        'new-activity-request', this._onNewActivity.bind(this)
+      );
 
-    if (typeof handler === 'function') {
-      this.setActivity(activity);
+      ActivityClient.on(
+        'share-activity-request', this._onShareActivity.bind(this)
+      );
 
-      handler.call(this, activity);
+      // Currently both client and service for activity reside in the same
+      // context, so we spin up both in the same context/window.
+      ActivityService.init();
```

Comment:
So it's not working if the client and service is in different context, right?  How about move this `ActivityService.init` in as a method of bridge and called in ActivityClient.init(but this means we'll need to spin up the service earlier and ActivityService.init only for register system message )?